File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/crypto/cipher/DecryptCipher.cpp
#include <base/buffer/util/BitsUtil.h>
#include <base/encoding/base64/Base64.h>
#include "DecryptCipher.h"
#include <base/MM.h>

using namespace base::crypto;

DecryptCipher::DecryptCipher(std::shared_ptr<IAlgorithm> algorithm)
    : super(algorithm)
{
    //void
}

int DecryptCipher::decryptBytesToInt(const char* eData) {
    char data[4];

    processBytes(eData, &data[0], 4);

    int ret = base::BitsUtil::getInt(&data[0], true);

    return ret;
}

std::string DecryptCipher::decryptBase64ToString(std::string& eData) {
    base::Buffer binaryEDataString = base::encoding::Base64::decode(eData);

    char* data = new char[binaryEDataString.getBytesCount()];

    processBytes(binaryEDataString.getBlockCharPointer(), &data[0], static_cast<int>(binaryEDataString.getBytesCount()));

    std::string ret{data, static_cast<size_t>(binaryEDataString.getBytesCount())};

    delete[] data;

    return ret;
}

DecryptCipher::~DecryptCipher() {
    //void
}

File: NodeJs_TypeScript_JavaScript_Server/Project_NodeJs_VSCode/src/base/crypto/cipher/DecryptCipher.ts
import { BasicCipher } from "./BasicCipher";
import { IDecryptCipher } from "./IDecryptCipher";
import { IAlgorithm } from "../algorithm/IAlgorithm";
import { addListener } from "cluster";
import { Base64 } from "base/encoding/base64/Base64";
import { LogicError } from "base/exception/LogicError";

export class DecryptCipher extends BasicCipher implements IDecryptCipher {

    public constructor(algorithm: IAlgorithm) {
        super(algorithm);

        //void
    }

    public decryptBase64ToString(eData: string): string {
        let binaryEDataString: Buffer = Base64.decode(eData);

        this.processBytes(
            binaryEDataString.buffer, binaryEDataString.byteOffset,
            binaryEDataString.buffer, binaryEDataString.byteOffset,
            binaryEDataString.byteLength
        );

        let ret: string = binaryEDataString.toString("utf8");
        
        return ret;
    }

    public decryptBufferToInt(eData: Buffer): number {
        if(eData == null || eData.length != 4) {
            throw new LogicError('Bounds error.');
        }

        let bData: Buffer = Buffer.allocUnsafe(4);
        
        this.processBytes(
            eData.buffer, eData.byteOffset,
            bData.buffer, bData.byteOffset,
            4
        );

        let data: number = bData.readUInt32BE(0);
        
        return data;
    }

}
